### DISCUSSION TITLE
reusing columns in multiple `Values` constructs

### DISCUSSION THREAD

**Discussion by aheld84** [sqlalchemy/sqlalchemy#10278] [Usage Questions] [answered]
Dear community,

I have a usage question regarding [`sqlalchemy.sql.expression.values`](https://docs.sqlalchemy.org/en/20/core/selectable.html#sqlalchemy.sql.expression.values):

When reusing the same columns in multiple `Values` constructs, the wrong data (always the data from the first construct) is used.

Here is a minimal example illustrating the issue (I'm using `Python 3.9` and `SQLAlchemy 2.0.20`):

```python
import sqlalchemy as sa

print("SQLAlchemy version:", sa.__version__)

columns = [sa.column("i", sa.Integer), sa.column("s", sa.String)]

for i in range(2):

    # if we put it here instead of outside the loop, everything works fine
    # columns = [sa.column("i", sa.Integer), sa.column("s", sa.String)]

    v = sa.values(
        *columns,
        name="data",
        literal_binds=True,
    ).data([(i, "foo"), (i + 10, "bar")])
    stmt = sa.select(v.c.i, v.c.s)
    print()
    print(f"{i=}")
    print(stmt.compile())

```

It produces this unexpected output:

```
SQLAlchemy version: 2.0.20

i=0
SELECT data.i, data.s 
FROM (VALUES (0, 'foo'), (10, 'bar')) AS data (i, s)

i=1
SELECT data.i, data.s 
FROM (VALUES (0, 'foo'), (10, 'bar')) AS data (i, s)
```

If I put the `columns` line inside the loop (see comment in the example), it produces the output I am expecting:

```
SQLAlchemy version: 2.0.20

i=0
SELECT data.i, data.s 
FROM (VALUES (0, 'foo'), (10, 'bar')) AS data (i, s)

i=1
SELECT data.i, data.s 
FROM (VALUES (1, 'foo'), (11, 'bar')) AS data (i, s)
```

From the [documentation](https://docs.sqlalchemy.org/en/20/core/selectable.html#sqlalchemy.sql.expression.values) I cannot tell whether I am using it wrong or this is a bug. Can someone please clarify this?

**Comment by zzzeek** [comment, marked as answer]
this is not the intended use because those `column` objects are bound to only one `Values` object at a time, and they apparently do not fully "re-bind" to a new `Values` construct; this case is not covered and it should raise an error.

**Reply by aheld84** [reply]
Thank you very much for the clarification. In my real use case, the columns are defined in a different part of the program and then used later for multiple `Values` constructs. I'm using `deepcopy` on the columns as a quick fix now but I will probably fix it by moving the definition of the columns to the part where the `Values` construct is used.

**Reply by zzzeek** [reply]
oh...no, please dont use deepcopy either, that's not supported either and can lead to problems.

use a function to create the columns in the first place and pass the function around.

**Reply by aheld84** [reply]
> use a function to create the columns in the first place and pass the function around

Yes, that is what I ended up doing. Deepcopy of the unbound columns was just a temporary work-around that happened to work in my case. But I understand that it is not supported in general.

**Reply by CaselIT** [reply]
@zzzeek is there a way of making things error when passed to copy/deepcopy? Since most sqlalchemy construct do not support it maybe it makes sense explicit warning/erroring?

**Reply by zzzeek** [reply]
you can implement `__copy__` and `__deepcopy__` to do something, but I think this would be very poorly received.    copy/deepcopy are of course very popular and people use them all over the place.

in the case with `column()` here, deepcopy probably does work as long as you have not yet used the column being copied in a fully compiled construct (that is, none of its internal memoizations are setup).  thats why the deepcopy thing is tricky.  it "works" in a lot of cases.

or we could actually go and make `__deepcopy__` actually work for SQL constructs too (selected ones, perhaps).   it just means lots of tests, having it warn / raise in cases where we really can't do a copy (like when you copy a table-bound column....did you want the table also?  what does that mean exactly?)

**Reply by CaselIT** [reply]
yes, that's more or less also my reasoning. let's leave as is for now

**Comment by zzzeek** [comment]
#10280 added, will raise in 2.0.21
